src/lem/macros.rs

```diff
@@ -0,0 +1,337 @@
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, let $tgt:ident : $tag:ident = hash4($src1:ident, $src2:ident, $src3:ident, $src4:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                $crate::lemop!(let $tgt: $tag = hash4($src1, $src2, $src3, $src4)),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, let ($tgt1:ident, $tgt2:ident) = unhash2($src:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                lemop!(let ($tgt1, $tgt2) = unhash2($src) ),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, let ($tgt1:ident, $tgt2:ident, $tgt3:ident) = unhash3($src:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                $crate::lemop!(let ($tgt1, $tgt2, $tgt3) = unhash3($src) ),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, let ($tgt1:ident, $tgt2:ident, $tgt3:ident, $tgt4:ident) = unhash4($src:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                lemop!(let ($tgt1, $tgt2, $tgt3, $tgt4) = unhash4($src) ),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, let $tgt:ident = hide($sec:ident, $src:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                $crate::lemop!(let $tgt = hide($sec, $src) ),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, let ($sec:ident, $src:ident) = open($hash:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                $crate::lemop!(let ($sec, $src) = open($hash) ),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, match_tag $sii:ident { $( $case:ident => $case_ops:tt ),* $(,)? } ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                $crate::lemop!( match_tag $sii { $( $case => $case_ops ),* } ),
+            },
+            $($tail)*
+        )
+    };
+    (@seq {$($limbs:tt)*}, return ($src1:ident, $src2:ident, $src3:ident) ; $($tail:tt)*) => {
+        $crate::lemop! (
+            @seq
+            {
+                $($limbs)*
+                $crate::lemop!( return ($src1, $src2, $src3) ),
+            },
+            $($tail)*
+        )
+    };
+}
+
+#[macro_export]
+macro_rules! lem {
+    ($in1:ident $in2:ident $in3:ident $lemop:tt) => {
+        $crate::lem::LEM::new(
+            [stringify!($in1), stringify!($in2), stringify!($in3)],
+            $crate::lemop!($lemop),
+        )
+    };
+}
```

Comment:
This is fine, though you could possibly use a different `@match` tag in the patterns above to merge with `lemop!`.

Reply:
I like this layering because it corresponds to the type hierarchy, in which a term of `LEMOP` is an attribute of a `LEM`